Text selection API conformance

jonleighton opened the pull request:
This is my next batch of changes for issue #19171. All the tests in tests/wpt/metadata/html/semantics/forms/textfieldselection/ are now passing (and also some tests outside of there).

I've made detailed notes about the changes in each commit message.

r? @KiChjang

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19544)
<!-- Reviewable:end -->

jonleighton committed:
Support the select() method on input/textarea

Issue #19171

jonleighton committed:
Fix selection{Start,End} when selectionDirection is "backward"

Per the spec, selectionStart and selectionEnd should return the same
values regardless of the selectionDirection. (That is, selectionStart is
always less than or equal to selectionEnd; the direction then implies
which of selectionStart or selectionEnd is the cursor position.)

There was no explicit WPT test for this, so I added one.

This bug was initially quite hard to wrap my head around, and I think
part of the problem is the code in TextInput. Therefore, in the process
of fixing it I have refactored the implementation of TextInput:

* Rename selection_begin to selection_origin. This value doesn't
  necessarily correspond directly to the selectionStart DOM value - in
  the case of a backward selection, it corresponds to selectionEnd.
  I feel that "origin" doesn't imply a specific ordering as strongly as
  "begin" (or "start" for that matter) does.

* In various other cases where "begin" is used as a synonym for "start",
  just use "start" for consistency.

* Implement selection_start() and selection_end() methods (and their
  _offset() variants) which directly correspond to their DOM
  equivalents.

* Rename other related methods to make them less wordy and more
  consistent / intention-revealing.

* Add assertions to assert_ok_selection() to ensure that our assumptions
  about the ordering of selection_origin and edit_point are met. This
  then revealed a bug in adjust_selection_for_horizontal_change() where
  the value of selection_direction was not maintained correctly (causing
  a unit test failure when the new assertion failed).

jonleighton committed:
Fix clearing the selection when value is changed

The implementation of adjust_horizontal_to_limit() is written with UI in
mind. As such, when there's a selection and we "adjust horizontal", the
selection will be cleared and the cursor will and up at the start/end of
the previous selection. This is what happens when you have a selection
and you press an arrow key on your keyboard, but it isn't the behaviour
we want when programmatically changing the value.

Instead, we need to first clear the selection, and then move the cursor
to the end. (We also need to reset the selection direction when clearing
the selection.)

jonleighton committed:
Don't fire select event when selection hasn't changed

jonleighton committed:
Implement setRangeText API

Spec: https://html.spec.whatwg.org/multipage/#dom-textarea/input-setrangetext

In order to do this, we need to define the SelectionMode enum in WebIDL:
https://html.spec.whatwg.org/multipage/#selectionmode

Since the enum is used by HTMLTextAreaElement and HTMLInputElement, it
doesn't seem to make sense to define it in the WebIDL file for one or
other of those.

However, we also can't create a stand-alone SelectionMode.webidl file,
because the current binding-generation code won't generate a "pub mod
SelectionMode;" line in mod.rs unless SelectionMode.webidl contains
either an interface or a namespace. (This logic happens in
components/script/dom/bindings/codegen/Configuration.py:35, in the
Configuration.__init__ method.)

I thought about changing the binding-generation code, but that seems
difficult. So I settled for placing the enum inside
HTMLFormElement.webidl, as that seems like a "neutral" location. We
could equally settle for putting it under HTMLTextAreaElement or
HTMLInputElement, it probably doesn't really matter.

The setRangeText algorithm set the "dirty value flag" on the
input/textarea. I made some clean-ups related to this:

1. HTMLTextAreaElement called its dirty value flag "value_changed"; I
   changed this to "value_dirty" to be consistent with the spec.

2. HTMLInputElement had a "value_changed" field and also a "value_dirty"
   field, which were each used in slightly different places (and
   sometimes in both places). I consolidated these into a single
   "value_dirty" field, which was necessary in order to make some of the
   tests pass.

TextControl::set_dom_range_text replaces part of the existing textinput
content with the replacement string (steps 9-10 of the algorithm). My
implementation changes the textinput's selection and then replaces the
selection. A downside of this approach is that we lose the original
selection state from before the call to setRangeText. Therefore, we have
to save the state into the original_selection_state variable so that we
can later pass it into TextControl::set_selection_range. This allows
TextControl::set_selection_range to correctly decide whether or not to
fire the select event.

An alternative approach would be to implement a method on TextInput
which allows a subtring of the content to be mutated, without touching
the current selection state. However, any such method would potentially
put the TextInput into an inconsistent state where the edit_point and/or
selection_origin is a TextPoint which doesn't exist in the content. It
would be up to the caller to subsequently make sure that the TextInput
gets put back into a valid state (which would actually happen, when
TextControl::set_selection_range is called).

I think TextInput's public API should not make it possible to put it
into an invalid state, as that would be a potential source of bugs.
That's why I didn't take this approach. (TextInput's public API does
currently make it possible to create an invalid state, but I'd like to
submit a follow-up patch to lock this down.)

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/htmltextareaelement.rs, components/script/dom/textcontrol.rs, components/script/dom/webidls/HTMLInputElement.webidl, components/script/dom/webidls/HTMLTextAreaElement.webidl, components/script/dom/htmlinputelement.rs and 2 more
 * @fitzgen: components/script/dom/htmltextareaelement.rs, components/script/dom/textcontrol.rs, components/script/dom/webidls/HTMLInputElement.webidl, components/script/dom/webidls/HTMLTextAreaElement.webidl, components/script/dom/htmlinputelement.rs and 2 more
 * @kichjang: components/script/dom/htmltextareaelement.rs, components/script/dom/textcontrol.rs, components/script/dom/webidls/HTMLInputElement.webidl, components/script/dom/webidls/HTMLTextAreaElement.webidl, components/script/dom/htmlinputelement.rs and 2 more

jonleighton committed:
Steps 7-9 of the input type change algorithm

Spec: https://html.spec.whatwg.org/multipage/input.html#input-type-change

In short, this resets the selection to the start of the field when the
type has changed from one which doesn't support the selection API to one
that does.

I couldn't see an existing WPT test covering this.

KiChjang reviewed (changes requested):
Not a full review, just commenting on things that jumped out at me when I skimmed through the changes.

jonleighton commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -687,6 +704,11 @@ impl HTMLInputElementMethods for HTMLInputElement {
 
+    // https://html.spec.whatwg.org/multipage/#dom-textarea/input-select
+    fn Select(&self) {
+        self.dom_select();
```
Thanks for starting to review this.

I can see where you are coming from, but personally I think that to try to put this code in `TextInput` would muddy the concerns of `TextInput`. At the moment it doesn't have any DOM node-related code - it is basically a wrapper around a string, providing some extra state about the selection and so on. This makes it easy to unit test and keeps its implementation focused. To put the selection API code inside `TextInput` would require passing in DOM nodes somehow.

As I say, I agree that there is some indirection here, I guess I just don't think it's that big a deal (easy to find the `dom_select` method using grep, ctags, etc...) It's also just how traits work - when you implement e.g. the `Iterator` trait, you get a whole load of extra methods which aren't defined in the same file, and that's the point of it.

There is potentially another option, though, which is to have a `TextControl` object rather than trait (maybe the name could be improved too). So `HTMLInputElement` and `HTMLTextAreaElement` would each do something like (in their constructor):

``` rust
self.text_control = TextControl::new(&self.upcast::<Node>(), self.textinput)
```

(Actually, we'd probably need to pass in more things than this... but you get the idea.)

And then this method would become:

``` rust
fn Select(&self) {
    self.text_control.dom_select();
}
```

Either way, since I'm not changing the existing architecture in this PR, could we perhaps move this discussion elsewhere, to an issue or something?

KiChjang replied:
At the very least, we'll need to make the code look less magical. I don't agree that "the ease of grepping for `dom_select`" is a good reason to not surprise future maintainers of the code, since even if I do find the definition in `text_control.rs`, it's not exactly apparent that both `HTMLInputElement` and `HTMLTextAreaElement` implement the `TextControl` trait, until you search both `htmlinputelement.rs` and `htmltextareaelement.rs` for its implementation.

Encapsulating a `TextInput` inside of a `TextControl` struct sounds fine to me, however. In other words, instead of having HTMLInputElement or HTMLTextAreaElement both contain a `TextInput` field, have a `TextControl` as a member field, inside of which contains a `TextInput`.

jonleighton replied:
Respectfully, I'm feeling quite frustrated that you're asking me to change this as a precondition of getting my code merged.

I had thought that the architectural question we're talking about had already been decided in #19272 (where you chose to defer the review to @jdm). I've now submitted this quite substantial PR and the only real feedback I've received after waiting several days is to ask me to make an architectural change which is unrelated to the actual PR, other than that it relates to the same area of the code.

I can understand your perspective, and I place a lot of value on easy to follow code too. I don't think either of us is "right", we just have a difference of opinion about the trade-offs of each choice. I'm willing to revisit the code that you're concerned about, but I'd prefer to do so in a separate PR, since this PR doesn't change the existing architecture. (And I think we should also involve @jdm in any such change since he reviewed #19272.)

KiChjang replied:
I'm sorry if my intentions were not clear! I did not mean that you need to make architectural changes in my previous comment. My simple wish is to keep the implementation details of `dom_select` within `htmlinputelement.rs` and `htmltextareaelement.rs`. A workable solution here is to not have a default implementation on the trait, and have it implemented on both `htmlinputelement.rs` and `htmltextareaelement.rs`.

Also, the architectural design is not already decided (refer to @jdm's comment [here](https://github.com/servo/servo/pull/19272#issuecomment-345870872)), but for the purpose of this PR, I concede that this is out of scope.

I really appreciate the work that you've done so far, and I understand it could be frustrating when a code review requires you to make huge architectural changes. In the future, may I invite you to #servo at `irc.mozilla.org` to discuss implementation plans, before any huge changes are made?

KiChjang replied:
Another idea would be to just leave a comment here explaining that this is implemented by default by the `TextControl` trait.

jonleighton replied:
> A workable solution here is to not have a default implementation on the trait, and have it implemented on both htmlinputelement.rs and htmltextareaelement.rs.

Not sure if I've misunderstood, but if you mean to duplicate the ~110 lines of code which implement `dom_select()` and `set_dom_range_text()` in `TextControl`, this seems like much more of a maintainability problem to me than a bit of indirection.

I'm fine with your other suggestion of adding a comment, though, so I've done that. I'm also willing to revisit the architecture of `TextControl` separately, as I already said.

> In the future, may I invite you to #servo at irc.mozilla.org to discuss implementation plans, before any huge changes are made?

I've been in the IRC room a bit, but I don't really understand what you'd prefer I'd done differently here? Do you mean that you'd prefer I'd discussed the #19272 changes in IRC before submitting a PR?

jonleighton commented:
@KiChjang I believe I've addressed everything you've currently commented on, would you mind having another look at this please? (Or let me know if I should ask somebody else?)

KiChjang commented on components/script/dom/textcontrol.rs:
```diff
@@ -3,18 +3,33 @@
     fn textinput(&self) -> &DomRefCell<TextInput<ScriptToConstellationChan>>;
     fn selection_api_applies(&self) -> bool;
+    fn has_selectable_text(&self) -> bool;
+    fn dirty_value_flag(&self) -> &Cell<bool>;
```
Returning a &Cell<bool> here looks like an implementation detail, rather than a generalized API. I would expect this to return a bool, and have another method that sets the dirty value flag.

KiChjang reviewed (changes requested):
Sorry for the delay! I've been stuck with reviewing the 2nd and the 2nd last commit forever, and as such I believe it's better to hand this review off to @nox, and perhaps @mbrubeck can also look at the 2nd commit.

KiChjang commented:
r? @nox and @mbrubeck (2nd commit)

jonleighton commented on components/script/dom/textcontrol.rs:
```diff
@@ -118,7 +122,116 @@ pub trait TextControl: DerivedFrom<EventTarget> + DerivedFrom<Node> {
+            },
+
+            SelectionMode::Preserve => {
+                // Step 1
```
How would you recommend labelling this? "Substep 1"? "Step 13 -> preserve -> 1"?

KiChjang replied:
I believe there are two ways to go about this, we can do `Substep 1` or `Step 13.1`.

jonleighton replied:
I'll go with the former; the latter seems ambiguous if other branches also had sub-steps.

mbrubeck reviewed (approved):
r+ for the second commit.  Looks great!

nox commented:
@KiChjang @mbrubeck reviewed the second commit, please approve your review and r=you if everything is ok.

jonleighton commented:
@nox FWIW I think @KiChjang was requesting that @mbrubeck review the 2nd commit only, and that you review the rest.

KiChjang commented:
@nox Sorry, I did mean to hand the PR over to you for review. Perhaps instead of the entire PR, you can look at the last 2 commits? r? @nox

jonleighton commented:
@KiChjang maybe we could just merge this? Seems like @nox is busy, and if there are no glaring problems with the code, surely having some implementation of this API is an improvement on having no implementation?

nox commented:
Great job with the last 2 commits! I think that can be merged as is.

@bors-servo r+

bors-servo commented:
:pushpin: Commit b76ec35 has been approved by `nox`

<!-- @bors-servo r=nox b76ec359d770f157926cbc04d9b67d5023cbe5c0 -->

bors-servo commented:
:hourglass: Testing commit b76ec359d770f157926cbc04d9b67d5023cbe5c0 with merge 462ac0d7bdedfcc16cbac5c352d22f72e01d085b...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7718)

nox commented:
There are a lot of new unexpected PASS.

jdm commented:
It should be enough to run `./mach test-wpt tests/wpt/web-platform-tests/dom/interfaces.html --log-raw wpt; ./mach update-wpt wpt`.

jonleighton commented:
@nox thanks for the review! I'm looking at fixing the build now

jonleighton commented:
@nox should be good to go now. (FWIW I had to add `--no-pause-after-test` to the `test-wpt` command; the file in question runs thousands of tests and it seemed to struggle with rendering all the output to the page...)

nox commented:
@bors-servo r+

Thanks again!

bors-servo commented:
:pushpin: Commit a8b64ac has been approved by `nox`

<!-- @bors-servo r=nox a8b64aca2a9c5e6e3756145afc0dedb606947ef8 -->

bors-servo commented:
:hourglass: Testing commit a8b64aca2a9c5e6e3756145afc0dedb606947ef8 with merge c2dfece49f1d59f51a3207cd3d88c282ee1adf70...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10546), [arm32](http://build.servo.org/builders/arm32/builds/10789), [arm64](http://build.servo.org/builders/arm64/builds/10818), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10839), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7705), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7722), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10915), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3313), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3336), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7579), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7650), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3410), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3357), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6221)
Approved by: nox
Pushing c2dfece49f1d59f51a3207cd3d88c282ee1adf70 to master...
